### ISSUE TITLE
Feature request: Skip children if parent fails

### ISSUE THREAD

**Issue by momchil-atanasov** [onsi/ginkgo#159]
Hi,

Here is an example that will make this easier to explain.

```go
Describe("IntegerFactory", func() {
  var factory IntegerFactory
  
  BeforeEach(func() {
    factory = NewIntegerFactory()
  })

  It("factory should not be nil", func() {
    Ω(factory).ShouldNot(BeNil())
  })

  Context("when two integers are created", func() {
    var first int
    var second int

    BeforeEach(func() {
      first = factory.CreateInt()
      second = factory.CreateInt()
    })

    It("integers should not be the same", func() {
      Ω(first).ShouldNot(Equal(second))
    })
  })
})
```

I believe that should `factory should not be nil` fail, then the `when two integers are created` branch should not be executed, since it will likely run on top of broken functionality and cannot be trusted, even if it were to pass. 

In our example case, if `NewIntegerFactory` returns `nil`, then it is certain that all the other tests will fail, resulting in unnecessary logs in the output and additional waiting.

Maybe there can be a flag in Ginkgo that will change the test execution flow so that child tests are executed only after successful parent tests.

**Upside:** It will make troubleshooting failing tests easier.

**Downside:** Child branches will always need to wait for all the tests in their parents to complete before they can be executed. This can have performance implications and could be difficult to implement.

**Comment by onsi** [comment]
Hey @momchil-atanasov 

In my mind this is somewhat equivalent to #70.  That thread captures my thoughts on why I don't think this sort of thing makes sense for Ginkgo.  A quick summary: it makes it hard to parallelize tests and to run tests in random order as it introduces an implicit relationship between different `It`s.

In general, I prefer making the assertion in my `BeforeEach`.  The `BeforeEach` is responsible for guaranteeing that the setup for its `Describe`/`Context` is correct.  This responsibility has two parts: performing the set up and (when relevant) validating it.  So, for your example:

```go
Describe("IntegerFactory", func() {
  var factory IntegerFactory
  
  BeforeEach(func() {
    factory = NewIntegerFactory()
    Ω(factory).ShouldNot(BeNil())
  })

  Context("when two integers are created", func() {
    var first int
    var second int

    BeforeEach(func() {
      first = factory.CreateInt()
      second = factory.CreateInt()
    })

    It("integers should not be the same", func() {
      Ω(first).ShouldNot(Equal(second))
    })
  })
})
```

Some people I've talked to think that making assertions in a `BeforeEach` is an anti-pattern.  I disagree.  Like I said - the `BeforeEach` is responsible for validating the tests.

Two annoying things come out of this pattern:

1. If the `BeforeEach` fails you get a lot of failures.  Yes this is noisy but, in practice, I find it fairly easy to diagnose and fix the problem immediately *precisely* because all the failures will be identical and will be coming from the same place: the assertion in the `BeforeEach`.

2. You sometimes have a *test* hiding in the `BeforeEach`.  For example, I deleted your `It` when I refactored your code.  In these circumstances I will sometimes actually include the original `It` purely for clarity.  An alternative would be to use `By` to call out that an interesting assertion is taking place:

```go
...
BeforeEach(func() {
  factory = NewIntegerFactory()
  By("creating a non-nil factory")
  Ω(factory).ShouldNot(BeNil())
})
...
```

**Comment by momchil-atanasov** [comment]
Since I wrote this issue, I've spent some time thinking about it and I have to agree that it would do more harm than good.

Still, I cannot fully agree that having the validation in the BeforeEach is the best way to go. In the example case above it does the job, but I can easily see case where it is unclear.

```go
Context("when an integer array is created with specific size", func() {
  var intArr IntegerArray
  
  BeforeEach(func() {
      intArr = factory.CreateIntArray(3)
  })
  
  It("should contain all integers to that size", func() {
    Ω(intArr.Size()).Should(Equal(3))
    Ω(intArr.Get(0)).ToEqual(1)
    Ω(intArr.Get(1)).ToEqual(2)
    Ω(intArr.Get(2)).ToEqual(3)
  })
  
  Context("when a value at some index is removed", func() {
    BeforeEach(func() {
      intArr.DeleteAt(1)
    })
    
    // Test something here
  })
})
```

Not having the `should contain all integers to that size` evaluation inside the BeforeEach code allows you to, at any given point in time, delete the `when a value at some index is removed` Context without having to redo any of the tests. In essence, the `when an integer array is created with specific size` Context is self-sufficient in the shown example. Whereas if you had all the It evaluations placed inside the BeforeEach, you would end up with a hollow Context.

Anyway, I guess we can have this issue closed. I just wanted to know whether there was something of the kind planned and what was your position on it. Seeing as it would introduce potential performance and randomization issues, I guess it's better to go without it.

**Comment by onsi** [comment]
Your example is interesting but I don't think I would advocate for moving everything in the `should contain all integers to that size` `It` to the `BeforeEach`.

I generally only include `nil`-checks and error validations in `BeforeEach`es.  *Maybe* you might include the `Ω(intArr.Size()).Should(Equal(3))` in the `BeforeEach` but even that is probably unnecessary.

**Comment by robdimsdale** [comment]
It sounds like we're all on the same page with regards to the value in having each `It` be independent of all others, as well as using `BeforeEach` mainly to assert that setup code was able to get the environment in the desired state, and that it did so without error, and not to assert behavior of the object under test.

@momchil-atanasov I'm going to close this issue. Please feel free to open another issue if you have further thoughts or questions.
